modules/ref_fb_module/src/fft_fb_impl.cpp

```diff
@@ -0,0 +1,234 @@
+        return;
+    }
+
+    try
+    {
+        if (inputDataDescriptor.getSampleType() == SampleType::Struct || inputDataDescriptor.getDimensions().getCount() > 0)
+            throw std::runtime_error("Incompatible input value data descriptor");
+
+        inputSampleType = inputDataDescriptor.getSampleType();
+        if (inputSampleType != SampleType::Float64 &&
+            inputSampleType != SampleType::Float32 &&
+            inputSampleType != SampleType::Int8 &&
+            inputSampleType != SampleType::Int16 &&
+            inputSampleType != SampleType::Int32 &&
+            inputSampleType != SampleType::Int64 &&
+            inputSampleType != SampleType::UInt8 &&
+            inputSampleType != SampleType::UInt16 &&
+            inputSampleType != SampleType::UInt32 &&
+            inputSampleType != SampleType::UInt64)
+            throw std::runtime_error("Invalid sample type");
+
+        if (inputDomainDataDescriptor.getSampleType() != SampleType::Int64 && inputDomainDataDescriptor.getSampleType() != SampleType::UInt64)
+        {
+            throw std::runtime_error("Incompatible domain data sample type");
+        }
+
+        const auto domainUnit = inputDomainDataDescriptor.getUnit();
+        if (domainUnit.getSymbol() != "s" && domainUnit.getSymbol() != "seconds")
+        {
+            throw std::runtime_error("Domain unit expected in seconds");
+        }
+
+        const auto domainRule = inputDomainDataDescriptor.getRule();
+        if (inputDomainDataDescriptor.getRule().getType() != DataRuleType::Linear)
+        {
+            throw std::runtime_error("FFT: Domain rule must be linear");
+        }
+
+        linearReader = BlockReaderFromExisting(linearReader, blockSize, SampleType::Float32, SampleType::UInt64);
+
+        auto dimensions = List<IDimension>();
+        dimensions.pushBack(Dimension(DimensionRuleBuilder()
+                                      .setType(DimensionRuleType::Linear)
+                                      .addParameter("delta", 1)
+                                      .addParameter("start", 1)
+                                      .addParameter("size", blockSize / 2)
+                                      .build()));
+
+        const auto inputRange = inputDataDescriptor.getValueRange();
+        outputDataDescriptor = DataDescriptorBuilder()
+                                   .setSampleType(SampleType::Float64)
+                                   .setDimensions(dimensions)
+                                   .setUnit(inputDataDescriptor.getUnit())
+                                   .setValueRange(inputRange.assigned() ? Range(0, inputRange.getHighValue()) : Range(0, 10))
+                                   .build();
+
+        outputSignal.setDescriptor(outputDataDescriptor);
+
+        outputDomainDataDescriptor =
+            DataDescriptorBuilderCopy(inputDomainDataDescriptor).setRule(ExplicitDataRule()).setSampleType(SampleType::UInt64).build();
```

Comment:
I guess the rule could be implicit here if you set it up correctly...

Reply:
Yes, but for now this is a bit simpler and the code is easier to understand. :) Can be updated in the future. In general this function block has a lot of possible optimizations.